### ISSUE TITLE
[Linux] Cannot demo heroes or play arcade gamemodes

### ISSUE THREAD

**Issue by rosshadden** [ValveSoftware/Dota-2#2290]
#### Your system information

* System information from steam (`Steam` -> `Help` -> `System Information`) in a [gist](https://gist.github.com/): https://gist.github.com/rosshadden/fba3b2e3378b01835af987f8d84672a0
* Have you checked for system updates?: Yes
* Are you using the latest stable video driver available for your system? Yes
* Have you verified the game files?: Yes

#### Please describe your issue in as much detail as possible:

For the past few months I have been unable to demo heroes or play any arcade gamemodes. I believe this started after some Dota update in November or December. I can play unranked all-pick, ranked all-pick, ability draft, and probably most other gamemodes that use the default map. But I am unable to demo heroes or play games from the arcade that I have tried. I am also unable to play tutorial scenarios like the List Hit Trainer. The only commonality I can think of is that these are all on non-default maps. I used to demo heroes or use last hit trainer pretty much daily on this computer and I believe this problem started occurring after a Dota or Steam update.

I have been using these launch params:

```
-novid -high -novr -nohltv -nojoy -prewarm -map dota -console -vulkan_disable_graphics_pipeline_library -vulkan_disable_pipeline_cache
```

However I have tried pretty much every permutation of these within reason, including removing them all. When I realized the default map coincidence I tried removing `-map dota` as well.

I'm not sure whether this is related, but with the dota files downloaded as-is from scratch, I am unable to launch the game. I have to replace `libtcmalloc_minimal.so.0` with my own `/usr/lib/libtcmalloc_minimal.so` in order to get it to launch. This has been brought up in other issues which is honestly what kept me from opening my own for several months. My issue seems so closely related to others that people have opened recently, but all of those have resulted in success for the poster and not for me.

Here is a crash report dump that Steam saved in `/tmp/dumps`: [crash.pcap.zip](https://github.com/ValveSoftware/Dota-2/files/11202890/crash.pcap.zip).
I will happily test or include anything else that would be helpful for troubleshooting!

#### Steps for reproducing this issue:

1. boot dota
2. demo a hero (or play last hit trainer, or play an arcade mode, etc)
3. game hangs and eventually crashes (or sometimes hangs indefinitely)

**Comment by kisak-valve** [comment]
For reference, the attached minidump is a SIGSEGV in libtcmalloc_minimal.so.0 coming from libscenefilecache.so.

**Comment by rosshadden** [comment]
Interesting... So I guess it is related then, and the workaround of using my system library is just masking/changing the problem. I also forgot to mention that I'm using steam-native but have also tried the normal self-contained runtime and still get the crash.

The `libtcmalloc_minimal.so` crash seems to be very commonly mentioned in issues here, but as far as I can recall every single resolution was to use the system library. I don't think I've seen a "real" solution to the problem. Maybe we can compile libtcmalloc ourselves from a specific version that is known to work?

**Comment by rosshadden** [comment]
I tried compiling and using `tcmalloc_minimal.so` from gperftools 2.9.1 and 2.8.1 (the latest version, which I have installed, is 2.10). I don't know the exact versioning of the tcmalloc libraries but I believe they were `4.5.9` and `4.5.6` respectively (the latest version and installed on my computer is `4.5.10`).
None of these attempts worked. I am still unable to demo any heroes, which I have been using as my main test case as has the shortest test cycle.

Edit: I have also tried `libtcmalloc_minimal_debug.so.4.5.10` to no avail. It seems to be functionally equivalent to the non-debug version though, as I am able to launch Dota and play normal games as expected.

Also, totally unrelated, my Steam username has ironically been `malloc` since before I started running into this issue, and it is indeed a reference to the C function which turned out to be a funny coincidence :rofl:

**Comment by rosshadden** [comment]
With the new patch coming out yesterday I just thought I should provide an update that I still have the same problem in 7.33. I can play normal matchmaking totally fine, but cannot demo heroes or play any custom game modes, last hit trainer, etc.

**Comment by stryan** [comment]
Try the lavapipe trick over in #2285 , that fixed it for me.

**Comment by rosshadden** [comment]
Good idea, but unfortunately didn't work for me. I tried both moving my lvp json file to disable it, and also launching steam with the `VK_LOADER_DRIVERS_DISABLE=*lvp*` env flag. I still cannot launch Dota with the libtcmalloc_minimal it ships with and instead need to use my own to get it to launch, and once launched the demo still does not work.

This is the error I get in the terminal when clicking "Demo hero", in steam's stdout:

```
memory allocation/deallocation mismatch at 0x555f505ca020: allocated with malloc being deallocated with delete
```

**Comment by rosshadden** [comment]
I got it working! It was a combination of a couple things. For one, I did in fact need to do the lavapipe trick @stryan suggestied. However, I also needed to let Steam use the libtcmalloc_minimal that Dota ships with. This was not something I had even thought to try because as stated before, I can't even _launch_ Dota with its version of tcmalloc. However if I launch Steam with lavapipe disabled it actually _does_ launch Dota with its own tcmalloc, and furthermore everything works. Thank you to all who helped me troubleshoot.
